drivers/stepper/gpio_stepper_controller.c

```diff
@@ -0,0 +1,228 @@
+						    STEPPER_SIGNAL_TARGET_POSITION_REACHED);
+			}
+		}
+	}
+}
+
+static int gpio_stepper_move(const struct device *dev, int32_t micro_steps,
+			     struct k_poll_signal *async)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+
+	K_SPINLOCK(&data->lock) {
+		if (data->async_signal) {
+			k_poll_signal_reset(data->async_signal);
+		}
+		data->async_signal = async;
+		data->step_count = micro_steps;
+		k_work_reschedule(&data->stepper_dwork, K_NO_WAIT);
+	}
+	return 0;
+}
+
+static int gpio_stepper_set_actual_position(const struct device *dev, int32_t position)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+
+	K_SPINLOCK(&data->lock) {
+		data->actual_position = position;
+	}
+	return 0;
+}
+
+static int gpio_stepper_get_actual_position(const struct device *dev, int32_t *position)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+
+	K_SPINLOCK(&data->lock) {
+		*position = data->actual_position;
+	}
+	return 0;
+}
+
+static int gpio_stepper_set_target_position(const struct device *dev, int32_t position,
+					    struct k_poll_signal *async)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+
+	K_SPINLOCK(&data->lock) {
+		if (data->async_signal) {
+			k_poll_signal_reset(data->async_signal);
+		}
+		data->async_signal = async;
+		data->step_count = position - data->actual_position;
+		k_work_reschedule(&data->stepper_dwork, K_NO_WAIT);
+	}
+	return 0;
+}
+
+static int gpio_stepper_is_moving(const struct device *dev, bool *is_moving)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+
+	*is_moving = k_work_delayable_is_pending(&data->stepper_dwork);
+	LOG_DBG("Is Motor Moving %d", *is_moving);
+	return 0;
+}
+
+static int gpio_stepper_set_max_velocity(const struct device *dev, uint32_t velocity)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+
+	if (velocity == 0) {
+		LOG_ERR("Velocity cannot be zero");
+		return -EINVAL;
+	}
+
+	K_SPINLOCK(&data->lock) {
+		data->delay_in_us = USEC_PER_SEC / velocity;
+	}
+	LOG_DBG("Setting Motor Speed to %d", velocity);
+	return 0;
+}
+
+static int gpio_stepper_get_micro_step_res(const struct device *dev,
+					   enum micro_step_resolution *micro_step_res)
+{
+	ARG_UNUSED(dev);
+	*micro_step_res = STEPPER_FULL_STEP;
+	return 0;
+}
+
+static int gpio_stepper_motor_controller_init(const struct device *dev)
+{
+	struct gpio_stepper_motor_controller_data *data = dev->data;
+	const struct gpio_stepper_motor_controller_config *config = dev->config;
+
+	data->dev = dev;
+	LOG_DBG("Initializing %s gpio_stepper_motor_controller", dev->name);
+	for (uint8_t n_pin = 0; n_pin < NUMBER_OF_GPIO_PINS; n_pin++) {
+		gpio_pin_configure_dt(&config->control_pins[n_pin], GPIO_OUTPUT_INACTIVE);
+	}
+	k_work_init_delayable(&data->stepper_dwork, stepper_work_step_handler);
+	return 0;
+}
+
+static const struct stepper_api gpio_stepper_api = {
+	.move = gpio_stepper_move,
+	.is_moving = gpio_stepper_is_moving,
```

Comment:
for what is this polling good? Does this mean that a upper lay have to ask again and again, if the motor is running? Maybe I stupid, but when a user tells move() he know it. Better is a sufficient event handling for FAULT detect delegation to upper layers and also reach a position or velocity, or ... whatever.